### PR TITLE
Fix name suffix not being applied when "patchesJson6902" is used

### LINKED ISSUE #4111: Kustomize doesn't append nameSuffix from overlay when patch is applied in patchesJson6902 section in version 4.2.0
nameSuffix isn't applied to serviceAccountName for daemonset, if patch is specified in `patchesJson6902` section, the same patch in section `patches` works fine.

See sample setup in https://github.com/Art3mK/kustomize-bug

**Expected output**

```yaml
➜  kustomize build overlays/prod
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::12345678910:role/fluentd-cw-logs
  name: fluentd-sa-abc
  namespace: abc
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: fluentd-abc
  namespace: abc
spec:
  template:
    spec:
      containers:
      - image: fluentd:latest
        name: fluentd
      serviceAccountName: fluentd-sa-abc
```

**Actual output**

```yaml
➜  kustomize build overlays/prod
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    eks.amazonaws.com/role-arn: arn:aws:iam::12345678910:role/fluentd-cw-logs
  name: fluentd-sa-abc
  namespace: abc
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: fluentd-abc
  namespace: abc
spec:
  template:
    spec:
      containers:
      - image: fluentd:latest
        name: fluentd
      serviceAccountName: fluentd-sa # <- missing nameSuffix
```

**Kustomize version**

{Version:kustomize/v4.2.0 GitCommit:d53a2ad45d04b0264bcee9e19879437d851cb778 BuildDate:2021-07-01T00:44:28+01:00 GoOs:darwin GoArch:amd64}

**Platform**

macOs

**Additional context**

adding random suffix to target in patch results in correct template
```yaml
➜  cat overlays/prod/kustomization.yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

bases:
  - "../../base"

namespace: abc
nameSuffix: -abc

# works fine with
#patches:
patchesJson6902:
  - path: patch-sa.yaml
    target:
      name: fluentd-sa-random # <- should give error, but it doesn't?
      kind: ServiceAccount
      version: v1
➜  kustomize build overlays/prod
apiVersion: v1
kind: ServiceAccount
metadata:
  name: fluentd-sa-abc
  namespace: abc
---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: fluentd-abc
  namespace: abc
spec:
  template:
    spec:
      containers:
      - image: fluentd:latest
        name: fluentd
      serviceAccountName: fluentd-sa-abc # <- lolwhut
```

### PR DESCRIPTION
**Pull request by Serializator**
Fixes #4111

### CODE REVIEW

File: api/internal/target/kusttarget_configplugin.go [hunk 1]
```diff
@@ -48,13 +48,13 @@ func (kt *KustTarget) configureBuiltinTransformers(
 	tc *builtinconfig.TransformerConfig) (
 	result []resmap.Transformer, err error) {
 	for _, bpt := range []builtinhelpers.BuiltinPluginType{
+		builtinhelpers.PatchJson6902Transformer,
 		builtinhelpers.PatchStrategicMergeTransformer,
 		builtinhelpers.PatchTransformer,
 		builtinhelpers.NamespaceTransformer,
```
Comment (natasha41575) [inline api/internal/target/kusttarget_configplugin.go:54]: While I agree that the `patchJson6902Transformer` should be placed next to the other patch transformers for consistency in behavior, I'm wondering if there might be any other unintended or possibly undesirable consequences for changing the order?

**Review by natasha41575 — COMMENTED** [review]
This PR needs a regression test, I suggest adding it to `api/krusty/multipatch_test.go`. You can look at the other tests in `api/krusty` for examples.

First commit should introduce the test with the wrong output demonstrating the failing behavior. The second commit should provide the fix, and correct the output in the test.

**Comment by Serializator** [comment]
> This PR needs a regression test, I suggest adding it to `api/krusty/multipatch_test.go`. You can look at the other tests in `api/krusty` for examples.

Excuse me, I completely neglected to run the tests locally, that would've made me aware of the regression before creating a pull request. I will check what and why it broke.

Using bisect I also pinpointed the commit in which it was broken (bd4580d73ab007b5efeb94460317e255a4cce44a), I will use that as reference point. It obviously isn't the order which broke it as this hasn't really changed since +/- 2019.

I will add the regression test once I pinpointed the root cause. Thank you for the feedback!

**Comment by Serializator** [comment]
I will communicate further status updates through the original issue (#4111). This way the scope of the conversation in the pull request is kept to the concrete implementation of the final solution rather than findings along the way.

**Comment by Serializator** [comment]
I pushed the proper solution I came up with after finding out this was caused by build annotations not being preserved after a JSON 6902 patch is applied.

I will a add the proper regression tests this evening or tomorrow.

@natasha41575, are files in `api/builtins/` generated automatically or should I run a command locally do to that? I'm not at all familiar with Pluginator and couldn't really find a helpful "quick start" documentation for now.

**Comment by natasha41575** [comment]
> are files in api/builtins/ generated automatically or should I run a command locally do to that? I'm not at all familiar with Pluginator and couldn't really find a helpful "quick start" documentation for now.

There is a command you can run, looking at the Makefile it might be `make generate-kustomize-builtin-plugins`? If that doesn't work I'm certain you can run `make prow-presubmit-check` and that will generate the code in `api/builtins` as well as run all the tests locally.

/ok-to-test

File: plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go [hunk 2]
```diff
@@ -6,6 +6,7 @@ package main
 
 import (
 	"fmt"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
```
Comment (natasha41575) [inline plugin/builtin/patchjson6902transformer/PatchJson6902Transformer.go:9]: nit: Move this to line 17 with the other `sigs.k8s.io` imports (and please make sure they are in alphabetical order)

File: plugin/builtin/patchtransformer/PatchTransformer.go [hunk 3]
```diff
@@ -6,6 +6,7 @@ package main
 
 import (
 	"fmt"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
```
Comment (natasha41575) [inline plugin/builtin/patchtransformer/PatchTransformer.go:9]: same comment here

**Review by natasha41575 — COMMENTED** [review]
LGTM apart from this extremely minor formatting nit. Thank you!

**Comment by natasha41575** [comment]
Also please take a look at [the CI errors](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kustomize/4266/kustomize-presubmit-master/1458534665445969920), it looks like there are some go-gets you have to run.

**Comment by Serializator** [comment]
@natasha41575, running `make prow-presubmit-check` locally to ensure that it will completely pass on my next commit. Resolved the nitpicks and will commit after all the tests pass! 👍🏼

**Comment by Serializator** [comment]
I don't understand why it is failing out of the sudden... @natasha41575, is it maybe because I didn't commit the `api/builtins` changes? Should I commit those or is that handled by the "kustomize-presubmit-master" job?

**Comment by natasha41575** [comment]
> I don't understand why it is failing out of the sudden... @natasha41575, is it maybe because I didn't commit the api/builtins changes? Should I commit those or is that handled by the "kustomize-presubmit-master" job?

Yes, please commit all your local changes.

**Comment by Serializator** [comment]
@natasha41575 I made the last changes and should pass all tests now. Sorry for the many back and forwards, still learning the workflow and how everything works / is structured within Kustomize 😄

**Comment by natasha41575** [comment]
/lgtm

Thanks for working on this!

**Pull request merged**